### PR TITLE
Classes cleanup and tests

### PR DESCRIPTION
**Pull request by lgiordani**
This is an implementation of #91 and #92 and partially #93.
(I opened this one to cleanup the history mess with #95)

### CODE REVIEW

File: bumpversion/engines.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+    def bump(self, value):
+        try:
+            return self._values[self._values.index(value)+1]
+        except IndexError:
+            raise ValueError(
+                "The part has already the maximum value among {} and cannot be bumped.".format(self._values))
+
+
+class DateEngine(object):
```
Comment (peritus) [inline bumpversion/engines.py]: This is new functionality, right ?

Comment (lgiordani) [inline bumpversion/engines.py]: Oh yes indeed, sorry I forgot to mention it

Comment (peritus) [inline bumpversion/engines.py]: If so, .. could you move that to it's own pull request so that we can discuss it individually (this needs docs (i know, "docs" is just "mentioned in readme" at this point) ?

That being said, I think (just trying to infer your intentions with this from the source code) the feature is already there. In the README there's this section:

> You can also use the variables now or utcnow to get a current timestamp. Both accept datetime formatting (when used like as in {now:%d.%m.%Y}).

.. how would that be different from the `DateEngine` ?

**Comment by peritus** [comment]
I like this pull request a lot in general.

I'm kind of hesitant on the naming of `*Engine` (with the meaning of "something that does something"). I would merge this regardless, but would probably search for a new name before making this part of a documented/public facing/stable/commited API. What would be a good name for these things ?

File: bumpversion/engines.py (hunk 1 above)
Comment (lgiordani) [inline bumpversion/engines.py]: Well... sorry but I completely missed the functionality that you quote from the README. I feel a bit stupid =) Given that, I still think it is a good thing to separate that behaviour in a dedicated engine, and will move this to a separate PR. This has to do with a future (but already ready to be committed) change, where a 'function' is explicitly associated to a version part, i.e. 'numeric', 'values', 'date'. We will discuss it on its own PR.

What is the best way to change the PR? Shall I just commit a remove of this class on the same branch?

Comment (lgiordani) [inline bumpversion/engines.py]: Ah but that functionality you mention is only for tags and commit message, isn't it? DateEngine is for a part of a version.

Comment (peritus) [inline bumpversion/engines.py]: I think the following should work:

```
git checkout classes_cleanup
# remove DateEngine with your favorite text editor
git add -u
git commit --amend
git push -f
```

Looking forward to your future plan :)

Comment (lgiordani) [inline bumpversion/engines.py]: Ok, I'll do it this afternoon. Thanks

Comment (peritus) [inline bumpversion/engines.py]: Yes, `now` being in the template context is kind of a hidden feature, there's lots to do in documentation / examples / cookbook ("how to use bumpversion if you want to use $PROGRAMMING_LANGUAGE to do $THING") / troubleshooting guidelines in this project.

So, you missing that might actually be a problem with the documentation, not you :)

Comment (peritus) [inline bumpversion/engines.py]: No hurry, I need to get back to paid work as well :) Thanks for everything!

**Comment by lgiordani** [comment]
Yes I was not really happy about *Engine too, but could not find another name at the time and then forgot it. Since I have to change the PR to move DateEngine I can also make a change here. Those *Engine are classes that rule the way a part of a version changes. So this is a "rule for a part of a version". "*Rule"? "*Function?" That upcoming next step I was talking about in a comment mkaes use of a configuration flag that I called "function", so perhaps "*Function" could be a more viable solution. What do you think?

**Comment by lgiordani** [comment]
I replaced `*Engine` with `*Function` and removed `DateEngine`.
